api/opensource/mfx_dispatch/linux/mfxloader.cpp

```diff
@@ -0,0 +1,583 @@
+    m_dlh = std::move(hdl);
+  } else {
+    m_uid = {};
+    m_create_plugin = nullptr;
+    m_plugin = {};
+    m_plugin_param = {};
+  }
+  return mfx_res;
+}
+
+mfxStatus PluginCtx::Unload()
+{
+  return MFXVideoUSER_Unregister((mfxSession)&m_loader, m_plugin_param.Type);
+}
+
+mfxStatus LoaderCtx::LoadPlugin(const mfxPluginUID &uid, mfxU32 version, const char *path)
+{
+  if (!path) return MFX_ERR_NULL_PTR;
+
+  try {
+    std::lock_guard<std::mutex> lock(m_guard);
+    for (auto it: m_plugins) {
+      if (it.getUID() == uid) return MFX_ERR_UNDEFINED_BEHAVIOR;
+    }
+  } catch(...) {
+    return MFX_ERR_UNKNOWN;
+  }
+
+  mfxStatus mfx_res = MFX_ERR_NONE;
+  std::list<PluginCtx> ctx;
+
+  try {
+    ctx.emplace_front(*this);
+  } catch(...) {
+    return MFX_ERR_UNKNOWN;
+  }
+
+  mfx_res = ctx.front().Load(uid, version, path);
+  if (MFX_ERR_NONE != mfx_res) {
+    return mfx_res;
+  }
+
+  try {
+    std::lock_guard<std::mutex> lock(m_guard);
+    m_plugins.splice(m_plugins.end(), ctx);
+  } catch(...) {
+    return MFX_ERR_UNKNOWN;
+  }
+
+  return MFX_ERR_NONE;
+}
+
+mfxStatus LoaderCtx::UnloadPlugin(const mfxPluginUID& uid)
+{
+  mfxStatus mfx_res = MFX_ERR_NONE;
+
+  try {
+    std::list<PluginCtx> ctx;
+    {
+      // We will move plugin ctx which we are going to delete to the
+      // array allocated on stack. In this way we will move bottom half
+      // of plugin ctx destroy, including potentially long dlclose, out
+      // of the mutex.
+      std::lock_guard<std::mutex> lock(m_guard);
+      auto it = std::find_if(std::begin(m_plugins), std::end(m_plugins),
+        [uid](const PluginCtx& cur_ctx){ return cur_ctx.getUID() == uid; });
+
+      if (it != std::end(m_plugins)) {
+          mfx_res = it->Unload();
+          if (MFX_ERR_NONE != mfx_res) {
+            return mfx_res;
+          }
+          ctx.splice(ctx.end(), m_plugins, it);
+      }
+    }
+  } catch(...) {
+    return MFX_ERR_UNKNOWN;
+  }
+
+  return MFX_ERR_NONE;
+}
+
+} // namespace MFX
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+mfxStatus MFXInit(mfxIMPL impl, mfxVersion *ver, mfxSession *session)
+{
+    mfxInitParam par{};
+
+    par.Implementation = impl;
+    if (ver) {
+        par.Version = *ver;
+    } else {
+        par.Version = VERSION(MFX_VERSION_MAJOR, MFX_VERSION_MINOR);
+    }
+
+    return MFXInitEx(par, session);
+}
+
+mfxStatus MFXInitEx(mfxInitParam par, mfxSession *session)
+{
+  if (!session) return MFX_ERR_NULL_PTR;
+
+  std::unique_ptr<MFX::LoaderCtx> loader;
+
+  try {
+    loader.reset(new MFX::LoaderCtx{});
+  } catch(...) {
+    return MFX_ERR_MEMORY_ALLOC;
+  }
+
+  mfxStatus mfx_res = loader->Init(par);
+  if (MFX_ERR_NONE == mfx_res) {
+    *session = (mfxSession)loader.release();
+  } else {
+    *session = nullptr;
+  }
+
+  return mfx_res;
+}
+
+mfxStatus MFXClose(mfxSession session)
+{
+  std::unique_ptr<MFX::LoaderCtx> loader((MFX::LoaderCtx*)session);
+
+  if (!loader) return MFX_ERR_INVALID_HANDLE;
```

Comment:
if you check the raw pointer before wrapping it to unique_ptr you will save one constructor call